lib/combineNodes.js

```diff
@@ -0,0 +1,310 @@
+var Cartesian3 = Cesium.Cartesian3;
+var combineMeshes = require('./combineMeshes');
+var combinePrimitives = require('./combinePrimitives');
+module.exports = combineNodes;
+
+//Combines nodes which are not targeted for animation and share the same material. 
+//Should be run after the combineMeshes and combinePrimitives stages, giving nodes one mesh each.
+function combineNodes(gltf, preserveCameras) {
+    var animations = gltf.animations;
+    var nodes = gltf.nodes;
+
+    //Give each node an extra property to check if they are not targeted for animation.
+    for (var animationId in animations) {
+        if (animations.hasOwnProperty(animationId)) {
+            var animation = animations[animationId];
+            if (defined(animation.channels)) {
+                var channelLength = animation.channels.length;
+                for (var i = 0; i < channelLength; i++) {
+                    nodes[animation.channels[i].target.id].extras._pipeline.isMergeable = false;
+                }
+            }
+        }
+    }
+    for (var nodeId in nodes) {
+        if (nodes.hasOwnProperty(nodeId)) {
+            var node = nodes[nodeId];
+            if (!defined(node.extras._pipeline.isMergeable)) {
+                node.extras._pipeline.isMergeable = true;
+            }
+        }
+    }
+
+    var scenes = gltf.scenes;
+    //Traverse the node tree and combine nodes from the bottom up.
+    if (defined(scenes)) {
+        for (var sceneId in scenes) {
+            if (scenes.hasOwnProperty(sceneId)) {
+                var roots = scenes[sceneId].nodes;
+                if (defined(roots)) {
+                    var rootsLength = roots.length;
+                    for (var j = 0; j < rootsLength; j++) {
+                        combineSubtree(gltf, roots[j], preserveCameras);
+                    }
+                }
+            }
+        }
+    }
+
+    //Combine the new meshes and primitives generated from this stage.
+    combineMeshes(gltf);
+    combinePrimitives(gltf);
+
+    return gltf;
+}
+
+//Recursively combines the children of a subtree from the bottom up.
+function combineSubtree(gltf, rootId, preserveCameras) {
+    var nodes = gltf.nodes;
+    var root = nodes[rootId];
+    var children = root.children;
+    if (defined(children) && children.length > 0) {
+        var childrenLength = children.length;
+        for (var i = 0; i < childrenLength; i++) {
+            combineSubtree(gltf, children[i], preserveCameras);
+        }
+
+        //Merge the children which are not animated and have the same properties into the current subtree root.
+        var childIndex = 0;
+        while (childIndex < children.length) {
+            var childId = children[childIndex];
+            var child = nodes[childId];
+            if (canCombine(root, child, preserveCameras)) {
+                mergeChildWithParent(gltf, rootId, childId, childIndex, preserveCameras);
+                children.splice(childIndex, 1);

```

Comment:
This can end of being O(n^2) since each splice O(n).  Perhaps just copy surviving children into a new array and assign it to `root.children`.